spec/unit_test_pr_checker_spec.rb

```diff
@@ -0,0 +1,167 @@
+# frozen_string_literal: true
+
+require_relative 'spec_helper'
+
+module Danger
+  describe Danger::UnitTestPRChecker do
+    it 'should be a plugin' do
+      expect(Danger::UnitTestPRChecker.new(nil)).to be_a Danger::Plugin
+    end
+
+    describe 'with Dangerfile' do
+      before do
+        @dangerfile = testing_dangerfile
+        @my_plugin = @dangerfile.unit_test_pr_checker
+
+        allow(@my_plugin.github).to receive(:pr_labels).and_return(['my_label'])
+      end
+
+      it 'shows that an Android PR needs tests' do
+        changes_dict = {
+          'File1.java' => 'import java.utils.*;\n\n public class Abc { public static void main(String[] args) { println(''); } }',
+          'project/src/androidTest/java/org/test/ToolTest.kt' => 'class ToolTest { void testMethod() {} }',
+          'File2.kt' => 'class Abcdef(name: String) { public void testMe() { println(''); } }'
+        }
```

Comment:
Those look like full file contents, as opposed to `diff` results like I'd expect from `git.diff`. I saw your implementation of `run_in_repo_with_diff` below but that means we'd be unable to test what happens if a class definition existed before the diff and was removed by the PR.

For example, imagine the case where before the PR we had a `class Foo` and tests for that `class Foo`. But the PR deletes the tests–while keeping the `class Foo` around. We want the plugin to detect that there's now no more tests for `Foo` and warn about it.

Or imagine that the PR removes both `class Foo` and its tests, we would want to make sure that the plugin see those as deletions, as opposed to accidentally (due to a flaw in our implementation, which is what our Unit Tests are checking for) see any changes as additions, even line removals.

Or finally, imagine if the definition of `class Foo` already existed before the PR, but the PR just moved that class definition into a new file (e.g. refactoring, or moving from a `.java` legacy code to `.kt` code…).

Those are all examples we need to test for, but your `run_in_repo_with_diff` helper would not be able to cover (and your current unit tests don't)

Reply:
> Those are all examples we need to test for, but your run_in_repo_with_diff helper would not be able to cover (and your current unit tests don't)

Agree -- I'm gonna refactor `run_in_repo_with_diff` and the tests to prefer mocking over the "real" git diffs to also help covering these additional cases.

Reply:
> For example, imagine the case where before the PR we had a class Foo and tests for that class Foo. But the PR deletes the tests–while keeping the class Foo around. We want the plugin to detect that there's now no more tests for Foo and warn about it.

Since this plugin is checking only the PR diff, I believe this case won't be covered by the nature of Danger?
Otherwise, I changed the tests a bit to cover other scenarios and added more test data.

But I was thinking we could eventually sophisticate this check with another plugin by using build artefacts and checking whether the test coverage has reduced with a new PR (which of course implies on saving the coverage state somewhere), having a similar effect as just checking the PR diff.